### PR TITLE
m_tx_download_mutex followups

### PR DESCRIPTION
**Pull request by glozow**
Followup to #30111. Includes suggestions:
- https://github.com/bitcoin/bitcoin/pull/30111#discussion_r1686303768
- https://github.com/bitcoin/bitcoin/pull/30111#discussion_r1686314984
- https://github.com/bitcoin/bitcoin/pull/30111#discussion_r1683186792
- https://github.com/bitcoin/bitcoin/pull/30111#issuecomment-2242819514
- https://github.com/bitcoin/bitcoin/pull/30111#discussion_r1686372826

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK.

File: src/net_processing.cpp [hunk 1]
```diff
@@ -6295,6 +6295,7 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
         //
         // Message: getdata (transactions)
         //
+        {
```
Comment (dergoegge) [inline src/net_processing.cpp:6300]: style nit: Indenting the body of the new scope would be nice

**Review by dergoegge — APPROVED** [review]
ACK 1a0212c8e5c25fce2e0a2cab3113b25f2fbc38b7

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2072,6 +2070,8 @@ void PeerManagerImpl::StartScheduledTasks(CScheduler& scheduler)
 
 void PeerManagerImpl::ActiveTipChange(const CBlockIndex* new_tip, bool is_ibd)
 {
+    // Ensure mempool mutex was released, otherwise deadlock may occur if another thread holding
+    // m_tx_download_mutex waits on the mempool mutex.
```
[the remark this answers is not in the document]
Comment (glozow) [reply src/net_processing.cpp:2074]: Hm, added a `ACQUIRED_BEFORE`, but compiler didn't complain when I added a `LOCK2(m_mempool.cs, m_tx_download_mutex)`

File: src/net_processing.cpp (hunk 1 above)
Comment (glozow) [reply src/net_processing.cpp:6300]: done

File: src/validationinterface.cpp [hunk 3]
```diff
@@ -183,6 +183,12 @@ void ValidationSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlo
                           fInitialDownload);
 }
 
+void ValidationSignals::ActiveTipChange(const CBlockIndex *new_tip, bool is_ibd)
+{
+    LOG_EVENT("%s: new block hash=%s block height=%d", __func__, new_tip->GetBlockHash().ToString(), new_tip->nHeight);
```
[the remark this answers is not in the document]
Comment (glozow) [reply src/validationinterface.cpp]: done

File: src/validation.cpp [hunk 4]
```diff
@@ -3553,7 +3553,7 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
             } // release MempoolMutex
             // Notify external listeners about the new tip, even if pindexFork == pindexNewTip.
             if (m_chainman.m_options.signals && this == &m_chainman.ActiveChainstate()) {
-                m_chainman.m_options.signals->ActiveTipChange(pindexNewTip, m_chainman.IsInitialBlockDownload());
+                m_chainman.m_options.signals->ActiveTipChange(*pindexNewTip, m_chainman.IsInitialBlockDownload());
```
Comment (instagibbs) [inline src/validation.cpp:3556]: aside: we could throw an `Assume()` on the pointer before deref, since it's not set in a non-trivial way above

**Review by instagibbs — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/30507/commits/1b732dbc3bd9c59303a774cf2095c2d6639dbe1d

File: src/validation.cpp (hunk 4 above)
Comment (dergoegge) [reply src/validation.cpp:3556]: (If you're gonna do this) `Assert` would make more sense, since we'd never want to hit the deref if it fails

**Review by theStack — APPROVED** [review]
lgtm ACK 1b732dbc3bd9c59303a774cf2095c2d6639dbe1d
(happy to re-ACK if the [`Assert` suggestion](https://github.com/bitcoin/bitcoin/pull/30507#discussion_r1689880324) for `pindexNewTip` is taken)

File: src/validation.cpp (hunk 4 above)
Comment (glozow) [reply src/validation.cpp:3556]: I could add `Assume` to the `if` condition?

**Comment by instagibbs** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/30507/commits/7c29e556c573a63351096c34bc072ae0c60ffa29

**Review by theStack — APPROVED** [review]
re-ACK 7c29e556c573a63351096c34bc072ae0c60ffa29

**Review by dergoegge — APPROVED** [review]
reACK 7c29e556c573a63351096c34bc072ae0c60ffa29

File: src/net_processing.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (hebasto) [reply src/net_processing.cpp:2074]: From https://clang.llvm.org/docs/ThreadSafetyAnalysis.html:
> `ACQUIRED_BEFORE(…)` and `ACQUIRED_AFTER(…)` are currently being developed under the `-Wthread-safety-beta` flag.

> Hm, added a `ACQUIRED_BEFORE`, but compiler didn't complain when I added a `LOCK2(m_mempool.cs, m_tx_download_mutex)`

With `CXXFLAGS=-Wthread-safety-beta`, clang emits a warning:
```
net_processing.cpp:2078:5: warning: mutex 'm_tx_download_mutex' must be acquired before 'cs' [-Wthread-safety-analysis]
    LOCK2(m_mempool.cs, m_tx_download_mutex);
    ^
./sync.h:260:16: note: expanded from macro 'LOCK2'
    UniqueLock criticalblock2(MaybeCheckNotHeld(cs2), #cs2, __FILE__, __LINE__)
               ^
1 warning generated.
```

**Review by hebasto — COMMENTED** [review]
Post-merge ACK 7c29e556c573a63351096c34bc072ae0c60ffa29.

**Review by Theghost256 — APPROVED** [review]

**Pull request merged**